### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by marianasvirzhevska**
JS Mentoring Homework 1

**### Task:**
**BE application setup**

> 1. Add index.ts file that just output “Hello, World!” into console and a test file with one dummy Jest test
> 2. Setup the following scripts for BE package:
> • build -> compile TypeScript
> • start -> run index.ts and watch it for changes (use tsc –w + nodemon)
> • lint -> run ESLint + Prettier check
> • test -> run the dummy test
> 3. Setup pre-commit hook to run lint and test scripts

**### FE application setup**

> 4. Add index.ts file that just output “Hello, World!” to the document body, index.scss with some styling and a test file with one dummy Jest test
> 5. Add needed loaders for your files
> 6. Setup HtmlWebpackPlugin
> 7. Setup Development configuration
> 8. Setup Production configuration
> 9. Setup the following scripts for FE package:
> a. build -> prepare production build
> b. start -> run dev server
> c. lint -> run ESLint + Prettier check
> d. test -> run the dummy test
> 10. Setup pre-commit hook to run lint and test scripts

### CODE REVIEW

File: api/__tests__/index.spec.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+describe('Start root file', () => {
+  test('Should print Hello World', () => {
+    // actual test
```
Comment (Panibrat) [inline api/__tests__/index.spec.ts]: Just write any dummy test

File: api/index.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import express = require('express');
+import { Request, Response } from 'express';
+
+const app = express();
+
+app.get('/', function (request: Request, response: Response) {
```
Comment (Panibrat) [inline api/index.ts]: No needs to implement API.  Task is: "Add index.ts file that just output “Hello, World!”

File: api/package.json [hunk 3]
```diff
@@ -0,0 +1,65 @@
+{
+  "name": "project-api",
+  "version": "1.0.0",
+  "description": "Simple node server example with typescript and pre-commit hooks.",
+  "main": "index.js",
+  "scripts": {
+    "start": "nodemon index.ts",
+    "test": "jest",
+    "lint": "eslint .",
+    "lint:fix": "eslint . --fix",
```
Comment (Panibrat) [inline api/package.json]: redundant

File: api/package.json [hunk 4]
```diff
@@ -0,0 +1,65 @@
+{
+  "name": "project-api",
+  "version": "1.0.0",
+  "description": "Simple node server example with typescript and pre-commit hooks.",
+  "main": "index.js",
+  "scripts": {
+    "start": "nodemon index.ts",
+    "test": "jest",
+    "lint": "eslint .",
```
Comment (Panibrat) [inline api/package.json:9]: This command should start ESLint and Prettier. Please follow the task.

File: api/package.json [hunk 5]
```diff
@@ -0,0 +1,65 @@
+{
+  "name": "project-api",
+  "version": "1.0.0",
+  "description": "Simple node server example with typescript and pre-commit hooks.",
+  "main": "index.js",
+  "scripts": {
+    "start": "nodemon index.ts",
```
Comment (Panibrat) [inline api/package.json]: It is better with a watcher mode: start -> run index.ts and watch it for changes (use tsc –w + nodemon)

File: app/src/index.html [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<!DOCTYPE html>
```
Comment (Panibrat) [inline app/src/index.html]: This file should be generated by HtmlWebpackPlugin plug-in

File: app/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+import './index.scss';
+
+window.onload = function() {
+  const root = document.getElementById('root');
+  root.innerText = 'Hello World!';
```
Comment (Panibrat) [inline app/src/index.ts]: Task is: Add index.ts file that just output “Hello, World!” to the document **body**

File: app/src/styles/fonts.scss [hunk 8]
```diff
@@ -0,0 +1,4 @@
+@font-face {
```
Comment (Panibrat) [inline app/src/styles/fonts.scss]: out of scope

File: app/webpack.config.js [hunk 9]
```diff
@@ -0,0 +1,142 @@
+const HtmlWebpackPlugin = require('html-webpack-plugin');
+const { CleanWebpackPlugin } = require('clean-webpack-plugin');
+const CopyWebpackPlugin = require('copy-webpack-plugin');
+const MiniCSSExtractPlugin = require('mini-css-extract-plugin');
+const OptimizeCssAssetsPlugin = require('optimize-css-assets-webpack-plugin');
+const TerserWebpackPlugin = require('terser-webpack-plugin');
+const ESLintPlugin = require('eslint-webpack-plugin');
+const path = require('path');
+
+const prod = process.env.NODE_ENV === 'production';
```
Comment (Panibrat) [inline app/webpack.config.js]: You can separate this file into two: one for dev, second for prod.

File: app/package.json [hunk 10]
```diff
@@ -0,0 +1,89 @@
+{
+  "name": "project-app",
+  "version": "1.0.0",
+  "description": "Simple application using typescript",
+  "main": "index.js",
+  "scripts": {
+    "test": "jest",
+    "lint": "eslint .",
```
Comment (Panibrat) [inline app/package.json:8]: This command should run ESLint + Prettier check

File: app/package.json [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    "lint:fix": "eslint . --fix",
```
Comment (Panibrat) [inline app/package.json]: redundant

File: app/package.json [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    "prettier": "prettier --write src/**/*.ts",
```
Comment (Panibrat) [inline app/package.json]: out of scope

File: app/package.json [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    "build": "cross-env NODE_ENV=production ./node_modules/.bin/webpack --mode production",
+    "start": "cross-env NODE_ENV=development ./node_modules/.bin/webpack serve --mode development",
+    "precommit": "lint-staged"
```
Comment (Panibrat) [inline app/package.json]: redundant

File: app/src/__tests__/index.spec.ts [hunk 14]
```diff
@@ -0,0 +1,5 @@
+describe('Start root file', () => {
+    test('Should print Hello World', () => {
+      // actual test
```
Comment (Panibrat) [inline app/src/__tests__/index.spec.ts]: Write a dummy test

File: app/src/index.scss [hunk 15]
```diff
@@ -0,0 +1,21 @@
+@import "~normalize.css";
```
Comment (Panibrat) [inline app/src/index.scss]: out of scope

File: .gitignore [hunk 16]
```diff
@@ -0,0 +1,17 @@
+coverage
+build
+node_modules
+
```
Comment (marianasvirzhevska) [inline .gitignore:4]: some comment test

File: api/package.json (hunk 4 above)
Comment (marianasvirzhevska) [reply api/package.json:9]: prettier runs from ESLint as a rule by "eslint-plugin-prettier" (it was set up in .eslintrc.json extends and plugins)

File: app/package.json (hunk 10 above)
Comment (marianasvirzhevska) [reply app/package.json:8]: prettier runs from ESLint as a rule by "eslint-plugin-prettier" (it was set up in .eslintrc.json extends and plugins)

File: api/package.json (hunk 5 above)
Comment (Panibrat) [reply api/package.json]: this command should start nodemon in "watch" mode.

Comment (marianasvirzhevska) [reply api/package.json]: Done in "MHW-1[Fixes] Resolve api issues" (see file changes)

**Pull request merged**